Plain highlighter throws UOE when using a script_score query wrapping an exists query

### Elasticsearch Version

8.10

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

unix

### Problem Description

If you try and highlight a `script_score` query that further wraps an `exists` query using the `plain` highlighter, you get a stack trace that looks like this:
```
Caused by: java.lang.UnsupportedOperationException
	at org.apache.lucene.search.highlight.WeightedSpanTermExtractor$DelegatingLeafReader.getFieldInfos(WeightedSpanTermExtractor.java:463) ~[lucene-highlighter-9.6.0.jar:?]
	at org.apache.lucene.search.FieldExistsQuery.rewrite(FieldExistsQuery.java:116) ~[lucene-core-9.6.0.jar:?]
	at org.apache.lucene.search.ConstantScoreQuery.rewrite(ConstantScoreQuery.java:44) ~[lucene-core-9.6.0.jar:?]
	at org.apache.lucene.search.ConstantScoreQuery.rewrite(ConstantScoreQuery.java:44) ~[lucene-core-9.6.0.jar:?]
	at org.apache.lucene.search.BooleanQuery.rewrite(BooleanQuery.java:289) ~[lucene-core-9.6.0.jar:?]
	at org.elasticsearch.common.lucene.search.function.ScriptScoreQuery.rewrite(ScriptScoreQuery.java:73) ~[elasticsearch-8.8.2.jar:?]
	at org.apache.lucene.search.highlight.WeightedSpanTermExtractor.extract(WeightedSpanTermExtractor.java:253) ~[lucene-highlighter-9.6.0.jar:?]
	at org.elasticsearch.search.fetch.subphase.highlight.CustomQueryScorer$CustomWeightedSpanTermExtractor.extract(CustomQueryScorer.java:77) ~[elasticsearch-8.8.2.jar:?]
	at org.apache.lucene.search.highlight.WeightedSpanTermExtractor.getWeightedSpanTerms(WeightedSpanTermExtractor.java:533) ~[lucene-highlighter-9.6.0.jar:?]
	at org.apache.lucene.search.highlight.QueryScorer.initExtractor(QueryScorer.java:209) ~[lucene-highlighter-9.6.0.jar:?]
	at org.apache.lucene.search.highlight.QueryScorer.init(QueryScorer.java:178) ~[lucene-highlighter-9.6.0.jar:?]
	at org.apache.lucene.search.highlight.Highlighter.getBestTextFragments(Highlighter.java:173) ~[lucene-highlighter-9.6.0.jar:?]
	at org.elasticsearch.search.fetch.subphase.highlight.PlainHighlighter.highlight(PlainHighlighter.java:151) ~[elasticsearch-8.8.2.jar:?]
	at org.elasticsearch.search.fetch.subphase.highlight.HighlightPhase$1.process(HighlightPhase.java:69) ~[elasticsearch-8.8.2.jar:?]

```

Related to #92919, but as `ScriptScoreQuery` is an elasticsearch query we need to extend `CustomWeightedSpanTermExtractor` to be aware of it.

### Steps to Reproduce

```
POST script-score/_doc
{
  "text": "some text"
}

GET script-score/_search
{
  "query": {
    "script_score": {
      "script": {
        "source": "_score + 1"
      },
      "query": {
        "bool": {
          "must": [
            {
              "match": {
                "text": "some"
              }
            },
            {
              "exists": { "field": "text" }
            }
          ]
        }
      }
    }
  },
  "highlight": {
    "type": "plain",
    "fields": {
      "*": {}
    }
  }
}
```

Returns

```
{
  "error": {
    "root_cause": [
      {
        "type": "unsupported_operation_exception",
        "reason": "unsupported_operation_exception: null"
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "script-score",
        "node": "TpI_NouPS1eIgeB4zNzGkg",
        "reason": {
          "type": "unsupported_operation_exception",
          "reason": "unsupported_operation_exception: null"
        }
      }
    ]
  },
  "status": 500
}
```

### Logs (if relevant)

_No response_

--- a/docs/changelog/99804.yaml
+++ b/docs/changelog/99804.yaml
@@ -1,0 +1,6 @@
+pr: 99804
+summary: Correctly handle `ScriptScoreQuery` in plain highlighter
+area: Highlighting
+type: bug
+issues:
+ - 99700

--- a/server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java
+++ b/server/src/main/java/org/elasticsearch/common/lucene/search/function/ScriptScoreQuery.java
@@ -67,6 +67,10 @@ public class ScriptScoreQuery extends Query {
         this.indexVersion = indexVersion;
     }
 
+    public Query getSubQuery() {
+        return subQuery;
+    }
+
     @Override
     public Query rewrite(IndexSearcher searcher) throws IOException {
         Query newQ = subQuery.rewrite(searcher);

--- a/server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java
+++ b/server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java
@@ -20,7 +20,7 @@ import org.elasticsearch.index.mapper.MapperBuilderContext;
 import org.elasticsearch.index.mapper.MapperService;
 import org.elasticsearch.index.mapper.MappingLookup;
 import org.elasticsearch.index.mapper.TextFieldMapper;
-import org.elasticsearch.script.ScriptService;
+import org.elasticsearch.script.ScriptCompiler;
 import org.elasticsearch.search.aggregations.support.ValuesSourceRegistry;
 import org.elasticsearch.xcontent.XContentParser;
 import org.elasticsearch.xcontent.XContentParserConfiguration;
@@ -50,7 +50,7 @@ public class QueryRewriteContext {
     protected final NamedWriteableRegistry writeableRegistry;
     protected final ValuesSourceRegistry valuesSourceRegistry;
     protected final BooleanSupplier allowExpensiveQueries;
-    protected final ScriptService scriptService;
+    protected final ScriptCompiler scriptService;
     private final XContentParserConfiguration parserConfiguration;
     protected final Client client;
     protected final LongSupplier nowInMillis;
@@ -73,7 +73,7 @@ public class QueryRewriteContext {
         final NamedWriteableRegistry namedWriteableRegistry,
         final ValuesSourceRegistry valuesSourceRegistry,
         final BooleanSupplier allowExpensiveQueries,
-        final ScriptService scriptService
+        final ScriptCompiler scriptService
     ) {
 
         this.parserConfiguration = parserConfiguration;

--- a/server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java
+++ b/server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java
@@ -48,9 +48,9 @@ import org.elasticsearch.index.mapper.SourceToParse;
 import org.elasticsearch.index.query.support.NestedScope;
 import org.elasticsearch.index.similarity.SimilarityService;
 import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptCompiler;
 import org.elasticsearch.script.ScriptContext;
 import org.elasticsearch.script.ScriptFactory;
-import org.elasticsearch.script.ScriptService;
 import org.elasticsearch.search.NestedDocuments;
 import org.elasticsearch.search.aggregations.support.ValuesSourceRegistry;
 import org.elasticsearch.search.lookup.LeafFieldLookupProvider;
@@ -110,7 +110,7 @@ public class SearchExecutionContext extends QueryRewriteContext {
         MapperService mapperService,
         MappingLookup mappingLookup,
         SimilarityService similarityService,
-        ScriptService scriptService,
+        ScriptCompiler scriptService,
         XContentParserConfiguration parserConfiguration,
         NamedWriteableRegistry namedWriteableRegistry,
         Client client,
@@ -183,7 +183,7 @@ public class SearchExecutionContext extends QueryRewriteContext {
         MapperService mapperService,
         MappingLookup mappingLookup,
         SimilarityService similarityService,
-        ScriptService scriptService,
+        ScriptCompiler scriptService,
         XContentParserConfiguration parserConfig,
         NamedWriteableRegistry namedWriteableRegistry,
         Client client,

--- a/server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/CustomQueryScorer.java
+++ b/server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/CustomQueryScorer.java
@@ -14,6 +14,7 @@ import org.apache.lucene.search.highlight.QueryScorer;
 import org.apache.lucene.search.highlight.WeightedSpanTerm;
 import org.apache.lucene.search.highlight.WeightedSpanTermExtractor;
 import org.elasticsearch.common.lucene.search.function.FunctionScoreQuery;
+import org.elasticsearch.common.lucene.search.function.ScriptScoreQuery;
 import org.elasticsearch.index.search.ESToParentBlockJoinQuery;
 
 import java.io.IOException;
@@ -73,6 +74,8 @@ public final class CustomQueryScorer extends QueryScorer {
                 super.extract(((FunctionScoreQuery) query).getSubQuery(), boost, terms);
             } else if (query instanceof ESToParentBlockJoinQuery) {
                 super.extract(((ESToParentBlockJoinQuery) query).getChildQuery(), boost, terms);
+            } else if (query instanceof ScriptScoreQuery ssq) {
+                super.extract(ssq.getSubQuery(), boost, terms);
             } else {
                 super.extract(query, boost, terms);
             }

--- a/server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighterTests.java
+++ b/server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/PlainHighlighterTests.java
@@ -16,9 +16,14 @@ import org.apache.lucene.tests.analysis.MockAnalyzer;
 import org.elasticsearch.index.mapper.MapperService;
 import org.elasticsearch.index.mapper.ParsedDocument;
 import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.script.DocReader;
+import org.elasticsearch.script.ScoreScript;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptContext;
 import org.elasticsearch.search.builder.SearchSourceBuilder;
 import org.elasticsearch.search.fetch.HighlighterTestCase;
 
+import java.io.IOException;
 import java.util.Map;
 
 public class PlainHighlighterTests extends HighlighterTestCase {
@@ -95,5 +100,51 @@ public class PlainHighlighterTests extends HighlighterTestCase {
             );
 
         }
+    }
+
+    public void testScriptScoreQuery() throws IOException {
+        MapperService mapperService = createMapperService("""
+            { "_doc" : { "properties" : {
+                "text" : { "type" : "text" }
+            }}}
+            """);
+
+        ParsedDocument doc = mapperService.documentMapper().parse(source("""
+            { "text" : "some text" }
+            """));
+
+        SearchSourceBuilder search = new SearchSourceBuilder().query(
+            QueryBuilders.scriptScoreQuery(
+                QueryBuilders.boolQuery().should(QueryBuilders.termQuery("text", "some")).must(QueryBuilders.existsQuery("text")),
+                new Script("foo")
+            )
+        ).highlighter(new HighlightBuilder().field("text").highlighterType("plain"));
+
+        assertHighlights(highlight(mapperService, doc, search), "text", "<em>some</em> text");
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    protected <T> T compileScript(Script script, ScriptContext<T> context) {
+        if (context == ScoreScript.CONTEXT) {
+            // Dummy script for testing ScriptScoreQuery highlighting
+            return (T) (ScoreScript.Factory) (params, lookup) -> new ScoreScript.LeafFactory() {
+                @Override
+                public boolean needs_score() {
+                    return true;
+                }
+
+                @Override
+                public ScoreScript newInstance(DocReader reader) throws IOException {
+                    return new ScoreScript(params, lookup, reader) {
+                        @Override
+                        public double execute(ExplanationHolder explanation) {
+                            return 0;
+                        }
+                    };
+                }
+            };
+        }
+        return super.compileScript(script, context);
     }
 }

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/MapperServiceTestCase.java
@@ -652,7 +652,7 @@ public abstract class MapperServiceTestCase extends ESTestCase {
             mapperService,
             mapperService.mappingLookup(),
             similarityService,
-            null,
+            MapperServiceTestCase.this::compileScript,
             parserConfig(),
             writableRegistry(),
             null,

